osquery/tables/applications/linux/namespace_ops.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <fcntl.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include <osquery/tables/applications/linux/namespace_ops.h>
+
+namespace osquery {
+namespace tables {
+
+Status NamespaceOps::invoke(std::function<void(int fd)> fn) {
+  std::string pidns = "/proc/" + std::to_string(_pid) + "/ns/mnt";
+  int fd = open(pidns.c_str(), O_RDONLY);
+  if (fd == -1) {
+    return Status(1, strerror(errno));
+  }
+
+  pid_t childPid = fork();
+  if (childPid == -1) {
+    int fErrno = errno;
+    close(fd);
+    return Status(1, strerror(fErrno));
+  }
+  if (childPid != 0) {
+    _childPid = childPid;
+    close(fd);
+    return Status();
+  }
+
+  // Change the namespace
+  if (setns(fd, 0) == -1) {
+    close(fd);
+    close(_sockFD);
+    _Exit(EXIT_SUCCESS);
+  }
+
+  // Execute the function
+  fn(_sockFD);
+
+  // Close the file handles and exit
+  close(fd);
```

Comment:
`close(2)` isn't guaranteed to succeed.  This needs to be wrapped in a `while` loop.  `close(2)` can be interrupted by signal delivery and can return non-0.

```
do {
  auto ret = close(fd);
} while(ret != 0);
```

This is one of my favorite scars from Linux that I will never forget.

Reply:
close() can fail with EINTR. What you are suggesting is not a good idea though. You are not checking if it failed with EINTR. So it can live lock if it failed for another reason.  Also, obviously the code won’t work. res is not accessible to the outer scope of while (). So unless there is another res symbol accessible in that while scope — in which case it’s a bug — compilation will fail.

Reply:
You actually *shouldn't* retry a close() which returns an error - the fd will still be closed regardless, the error is usually a deferred failure from a previous IO. In fact the manpage for close(2) spells this out in detail:

> Not  checking  the return value of close() is a common but nevertheless serious programming error.  It is quite possible that errors on a previous write(2) operation are first reported at the final close().  Not checking the return value when closing the file may lead to silent loss of data.  This can especially be observed with NFS and with disk quota.  *Note that the return value should be used only  for  diagnostics.* In  particular close() should not be retried after an EINTR since this may cause a reused descriptor from another thread to be closed.

So, the return value must be checked for diagnostic purposes, to reliably detect IO failures, but should not be treated as a retryable condition.

Reply:
So I am guessing the action item for me here is to log any errors from close.